source/navier_stokes_matrix.cc

```diff
@@ -848,45 +847,60 @@ NavierStokesMatrix<dim>::local_operation(
-                sym          = tmu * (grad_u[1][2] + grad_u[2][1]);
-                grad_u[1][2] = sym;
-                grad_u[2][1] = sym;
-              // fall through to part that is also present also in 2D
-              case 2:
-                sym          = tmu * (grad_u[0][1] + grad_u[1][0]);
-                grad_u[0][1] = sym;
-                grad_u[1][0] = sym;
-                break;
-              case 1: // nothing todo
-                break;
-              default:
-                Assert(false, ExcNotImplemented());
+                  case 3:
+                    sym          = tmu * (grad_u[0][2] + grad_u[2][0]);
+                    grad_u[0][2] = sym;
+                    grad_u[2][0] = sym;
+                    sym          = tmu * (grad_u[1][2] + grad_u[2][1]);
+                    grad_u[1][2] = sym;
+                    grad_u[2][1] = sym;
+                  // fall through to part that is also present also in 2D
+                  case 2:
+                    sym          = tmu * (grad_u[0][1] + grad_u[1][0]);
+                    grad_u[0][1] = sym;
+                    grad_u[1][0] = sym;
+                    break;
+                  case 1: // nothing todo
+                    break;
+                  default:
+                    Assert(false, ExcNotImplemented());
 #pragma GCC diagnostic push
-            }
+                }
 
-          // add pressure
-          for (unsigned int d = 0; d < dim; ++d)
+              for (unsigned int d = 0; d < dim; ++d)
+                {
+                  grad_u[d][d] =
+                    tmu_times_2 * grad_u[d][d] + parameters.tau_grad_div * divergence;
+
+                  // In case of Newtonian compressible constitutive type, subtract the
+                  // volumetric part from the rate of deformation tensor.
+                  if (parameters.constitutive_type ==
+                      FlowParameters::newtonian_compressible_stokes_hypothesis)
+                    grad_u[d][d] -= tmu_times_2 * divergence / static_cast<double>(dim);
+                }
+            }
+          else if (parameters.constitutive_type == FlowParameters::user_defined)
             {
-              grad_u[d][d] =
-                tmu_times_2 * grad_u[d][d] + parameters.tau_grad_div * divergence;
-
-              // Subtract the volumetric part from the rate of deformation tensor.
-              if (parameters.constitutive_type ==
-                  FlowParameters::newtonian_compressible_stokes_hypothesis)
-                grad_u[d][d] -= tmu_times_2 * divergence / static_cast<double>(dim);
-
-              if (LocalOps == NavierStokesOps::vmult ||
-                  LocalOps == NavierStokesOps::residual)
-                grad_u[d][d] -= pres;
+              Tensor<2, dim, vector_t> grad_u_temp =
+                convert_to_tensor<2, dim, vector_t>(velocity.get_gradient(q));
+              grad_u =
+                tau1 * user_defined_material(grad_u_temp, cell, q, q == 0 /*do_reinit*/);
```

Comment:
How is this supposed to be used? `do_reinit` indicates that the user might be using `FEvaluation` in the function? In that case you have to know fe_index/quad_index there. Is that known there?

Reply:
I'll provide a clearer description and exemplary usage of the interface in the evening.